Kevin Cheng
Yuting Huang

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
  

file = "CTG.csv"
# Read the CSV file.
data = pd.read_csv(file, skiprows=1)


# Select the relevant numerical columns.
selected_cols = ['LB', 'AC', 'FM', 'UC', 'DL', 'DS', 'DP', 'ASTV', 'MSTV', 'ALTV',
                 'MLTV', 'Width', 'Min', 'Max', 'Nmax', 'Nzeros', 'Mode', 'Mean',
                 'Median', 'Variance', 'Tendency', 'NSP']
data = data[selected_cols].dropna()

# Shuffle the dataset.
data_shuffled = data.sample(frac=1.0, random_state=0)

# Split into input part X and output part Y.
X = data_shuffled.drop('NSP', axis=1)

# Map the diagnosis code to a human-readable label.
def to_label(y):
    return [None, 'normal', 'suspect', 'pathologic'][(int(y))]

Y = data_shuffled['NSP'].apply(to_label)

# Partition the data into training and test sets.
Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.2, random_state=0)

In [3]:
X.head()

LB   AC   FM   UC   DL   DS   DP  ASTV  MSTV  ALTV  ...  Width  \
658   130.0  1.0  0.0  3.0  0.0  0.0  0.0  24.0   1.2  12.0  ...   35.0   
1734  134.0  9.0  1.0  8.0  5.0  0.0  0.0  59.0   1.2   0.0  ...  109.0   
1226  125.0  1.0  0.0  4.0  0.0  0.0  0.0  43.0   0.7  31.0  ...   21.0   
1808  143.0  0.0  0.0  1.0  0.0  0.0  0.0  69.0   0.3   6.0  ...   27.0   
825   152.0  0.0  0.0  4.0  0.0  0.0  0.0  62.0   0.4  59.0  ...   25.0   

        Min    Max  Nmax  Nzeros   Mode   Mean  Median  Variance  Tendency  
658   120.0  155.0   1.0     0.0  134.0  133.0   135.0       1.0       0.0  
1734   80.0  189.0   6.0     0.0  150.0  146.0   150.0      33.0       0.0  
1226  120.0  141.0   0.0     0.0  131.0  130.0   132.0       1.0       0.0  
1808  132.0  159.0   1.0     0.0  145.0  144.0   146.0       1.0       0.0  
825   136.0  161.0   0.0     0.0  159.0  156.0   158.0       1.0       1.0  

[5 rows x 21 columns]

In [4]:
from sklearn.dummy import DummyClassifier

clf = DummyClassifier(strategy='most_frequent')

In [5]:
from sklearn.model_selection import cross_val_score

cross_val_score(clf, Xtrain, Ytrain)

array([0.78235294, 0.78235294, 0.77941176, 0.77941176, 0.77941176])

In [6]:
# step 3 - train different classifiers

# Decision Tree
from sklearn.tree import DecisionTreeClassifier 

clf1 = DecisionTreeClassifier(max_depth=4) # prevent overfitting

cross_val_score(clf1, Xtrain, Ytrain)

array([0.92352941, 0.94117647, 0.91176471, 0.90882353, 0.92058824])

In [7]:
# Random Forest 
from sklearn.ensemble import RandomForestClassifier

clf2 = RandomForestClassifier()

cross_val_score(clf2, Xtrain, Ytrain)

array([0.93529412, 0.95294118, 0.93235294, 0.92941176, 0.95      ])

In [8]:
# LinearSVC
from sklearn.svm import LinearSVC 

clf3 = LinearSVC()

cross_val_score(clf3, Xtrain, Ytrain)

array([0.88823529, 0.92058824, 0.86764706, 0.89411765, 0.90588235])

In [9]:
# MLP Classifier
from sklearn.neural_network import MLPClassifier  

clf4 = MLPClassifier()

cross_val_score(clf4, Xtrain, Ytrain)

array([0.86764706, 0.90294118, 0.87058824, 0.90588235, 0.88823529])

In [10]:
from sklearn.metrics import accuracy_score
clf2.fit(Xtrain, Ytrain) # using random forest
Yguess = clf2.predict(Xtest)
print(accuracy_score(Ytest, Yguess))

0.9342723004694836


Decision tree classifier
array([0.92352941, 0.94117647, 0.91176471, 0.90882353, 0.92058824])


Randomforest classifier
array([0.93529412, 0.95294118, 0.93235294, 0.92941176, 0.95      ])

LinearSVC
array([0.88823529, 0.92058824, 0.86764706, 0.89411765, 0.90588235])

